crypto/rsa_extra/rsassa_pss_asn1_test.cc

```diff
@@ -0,0 +1,367 @@
+     NID_mgf1, NID_sha256, omit_salt_len},
+    {jdk_pss_sha256_salt_30, sizeof(jdk_pss_sha256_salt_30), NID_sha256,
+     NID_mgf1, NID_sha256, 30},
+    {jdk_pss_sha224_salt_absent, sizeof(jdk_pss_sha224_salt_absent), NID_sha224,
+     NID_mgf1, NID_sha224, omit_salt_len},
+    {jdk_pss_sha384_salt_absent, sizeof(jdk_pss_sha384_salt_absent), NID_sha384,
+     NID_mgf1, NID_sha384, omit_salt_len},
+    {jdk_pss_sha512_salt_absent, sizeof(jdk_pss_sha512_salt_absent), NID_sha512,
+     NID_mgf1, NID_sha512, omit_salt_len},
+    {jdk_pss_sha256_mgf1_sha512, sizeof(jdk_pss_sha256_mgf1_sha512), NID_sha256,
+     NID_mgf1, NID_sha512, omit_salt_len},
+    {jdk_pss_sha1_mgf1_sha256, sizeof(jdk_pss_sha1_mgf1_sha256), NID_undef,
+     NID_mgf1, NID_sha256, omit_salt_len},
+    {jdk_pss_sha256_mgf1_sha1, sizeof(jdk_pss_sha256_mgf1_sha1), NID_sha256,
+     NID_undef, NID_undef, omit_salt_len},
+    {jdk_pss_sha1_mgf1_sha1_salt_30, sizeof(jdk_pss_sha1_mgf1_sha1_salt_30), NID_undef,
+     NID_undef, NID_undef, 30},
+};
+
+class RsassaPssTest: public testing::TestWithParam<PssParamsTestInput> {};
+
+TEST_P(RsassaPssTest, DecodeParamsDer) {
+  const auto &param = GetParam();
+  CBS params;
+  params.data = param.der;
+  params.len = param.der_len;
+  RSASSA_PSS_PARAMS *pss = nullptr;
+  ASSERT_TRUE(RSASSA_PSS_parse_params(&params, &pss));
+  // Holds ownership of heap-allocated RSASSA_PSS_PARAMS.
+  bssl::UniquePtr<RSASSA_PSS_PARAMS> pss_ptr(pss);
+  // Expect all bytes of params are used.
+  ASSERT_FALSE(CBS_len(&params));
+  // Validate Hash Algorithm of RSASSA-PSS-params.
+  if (param.expected_hash_nid) {
+    ASSERT_TRUE(pss->hash_algor);
+    EXPECT_EQ(pss->hash_algor->nid, param.expected_hash_nid);
+  } else {
+    // When no expectation, make sure this value is absent.
+    EXPECT_FALSE(pss->hash_algor);
+  }
+  // Validate Mask Algorithm of RSASSA-PSS-params.
+  RSA_MGA_IDENTIFIER *mga = pss->mask_gen_algor;
+  if (param.expected_mask_gen_nid) {
+    ASSERT_TRUE(mga);
+    ASSERT_TRUE(mga->mask_gen);
+    EXPECT_EQ(mga->mask_gen->nid, param.expected_mask_gen_nid);
+    // Validate Hash Algorithm of Mask Gen.
+    if (param.expected_mgf1_hash_nid) {
+      ASSERT_TRUE(mga->one_way_hash);
+      EXPECT_EQ(mga->one_way_hash->nid, param.expected_mgf1_hash_nid);
+    } else {
+      // When no expectation, make sure this value is absent.
+      EXPECT_FALSE(mga->one_way_hash);
+    }
+  } else {
+    // When no expectation, make sure this value is absent.
+    EXPECT_FALSE(mga);
+  }
+  // Validate Minimum Salt Length of RSASSA-PSS-params.
+  if (param.expected_salt_len != omit_salt_len) {
+    ASSERT_TRUE(pss->salt_len);
+    EXPECT_EQ(pss->salt_len->value, param.expected_salt_len);
+  } else {
+    // When no expectation, make sure this value is absent.
+    EXPECT_FALSE(pss->salt_len);
+  }
+  // Perform signature generation MUST omit the trailerField field.
+  // https://tools.ietf.org/html/rfc4055#page-8
+  EXPECT_FALSE(pss->trailer_field);
```

Comment:
We should still have a unit test that shows we can parse it correctly?

Reply:
Discussed in another thread.